byway/src/main/java/com/google/sps/servlets/StopsServlet.java

```diff
@@ -26,26 +36,35 @@
 /** Servlet that handles data for stops */
 @WebServlet("/api/stop")
 public final class StopsServlet extends HttpServlet {
-  @GuardedBy("this")
-  private ArrayList<String> stops = new ArrayList<String>();
-
   private final Gson gson = new Gson();
-  // Implement datastore in routepage branch
+  private final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+  private static final Type ARRAYLIST_STRING = new TypeToken<ArrayList<String>>() {}.getType();
+  private static final FluentLogger logger = FluentLogger.forEnclosingClass();
 
+  /* Passes saved destinations stops (if any) to be shown in the schedule panel */
   @Override
-  public synchronized void doGet(HttpServletRequest request, HttpServletResponse response)
-      throws IOException {
-    Gson gson = new Gson();
-    String json = gson.toJson(stops);
-
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    List<String> stops = Collections.emptyList();
+    try {
+      Entity entity = datastore.get(KeyFactory.createKey(Trip.KIND, 5910974510923776L));
+      stops = (ArrayList<String>) entity.getProperty("destinations");
+    } catch (EntityNotFoundException e) {
+      logger.atInfo().withCause(e).log("Could not retrieve stops due to faulty key", e);
```

Comment:
Note that you have no `%s` formatting parameters in the logged message, so the `e` being passed at the end won't actually be formatted into the message. Also note that you're already using `withCause(e)` which will appropriately include the cause in the error, so you don't really need to format the error into the message string.

For the message string, try to be more specific, e.g. if reading this message without being familiar with this code what does "faulty key" mean? Consider something like `.log("Could not retrieve Entity for Trip with key %s while trying to get the stops", key)` (note that to do this you need to store the result of `KeyFactory.createKey` before the `try` block). That way if this error happens, you can definitely see what key was *actually* provided and have an easier time debugging. This applies to other places you have log messages as well. Be as specific as you reasonably can, so it's easy to tell:

- What happened ("Could not retrieve Entity for Trip")
- What data it affected ("with key %s")
- What you were trying to do ("while trying to get the stops")

This kind of context can make it a lot easier to tell why something caused an issue, or even make it possible to tell right away when what you actually did doesn't match what you were trying to do; for example, if `doPost` says "while trying to *update* the stops", you can tell immediately whether the error was in the post request or the get request.

Reply:
I didn't notice I took out the format param and key when I switched to logger. I will try to be more detailed with my error messages, so thanks for breaking down what exactly needs to go into the error message.